### PR TITLE
Upgrade to Catch2 version 3

### PR DESCRIPTION
**Pull request by nickbianco**
Partially motivated by #3616.

### Brief summary of changes
This PR upgrades the Catch2 testing library we use to version 3. The biggest change is that Catch2 is no longer installed via a `catch.hpp` file, but rather compiled from source into a static library and installed as a proper dependency. More details about [migrating to version 3 can be found here](https://github.com/catchorg/Catch2/blob/devel/docs/migrate-v2-to-v3.md). 

A few details about this change:
- The CMake files for dependencies have been updated to include Catch2 using the target `catch2`. I'm using the latest Catch2 version, 3.5.0.
- All includes now use `#include <catch2/catch_all.hpp>`.
- `#define CATCH_CONFIG_MAIN` is no longer needed.
- Tests have been updated to use the correct namespaces (e.g., `Catch::Approx`).
- `Catch::Contains` is now `Catch::Matchers::ContainsSubstring`.
- One drive-by: avoided unused variable warnings in `testMuscles.cpp` and `testProbes.cpp`.

### Testing I've completed

All of it, I guess? ([Who tests the testers?](https://en.wikipedia.org/wiki/Quis_custodiet_ipsos_custodes%3F))

### Looking for feedback on...

One feature of Catch2 version 3 is that the testing libraries can be included modularly. Currently, I just include everything in all tests (i.e., `catch_all.hpp`), but perhaps including only what is needed in each could be more efficient. 

### CHANGELOG.md (choose one)

- no need to update because...for internal testing.

- - -
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3655)

### CODE REVIEW

**Comment by nickbianco** [comment]
When/if this is merged, we should update the instructions in #3555.

**Review by adamkewley — COMMENTED** [review]
Nice changes, good work

I'd strongly recommend figuring out how to handle test dependencies separately from runtime dependencies. The current way you've written this forces all downstream library consumers to eat Catch - even if they don't plan on running the test suite or using Catch themselves.

It's problematic not only because it's unnecessary, but because it will cause issues in downstream codebases that use their own independent copy of Catch: even if they link their copy of Catch the "correct" way, they'll still run into problems related to either ambiguous includes or multiple symbol definitions, or (worst) runtime link ambiguities.

As for *how* to do this, it usually involves going into the test suite(s) and ensuring they separately link to the appropriate test dependency. E.g. OSC is handling it like this:

- https://github.com/ComputationalBiomechanicsLab/opensim-creator/blob/0.5.5/tests/testoscar/CMakeLists.txt#L61

**Review by nickbianco — COMMENTED** [review]
@adamkewley, that makes a lot of sense! It was easy to move all Catch2 linkages to the test suites. Ready for another review.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3655#-:-Nm1tzs1Ees8ToZYpyTT:b-lwki17)* status: 0 of 69 files reviewed, 11 unresolved discussions (waiting on @adamkewley)

___
*[`CMakeLists.txt` line 257 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431075522:-Nm1p-EwELWduklfrm8b:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/CMakeLists.txt#L257)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

It feels a little weird that `catch` is listed as an *install* dependency (as in, anyone downstream who wants to use OpenSim will now also get Catch). Anecdotally, I'd say that most projects put test dependencies separately from runtime ones
</blockquote></details>

Done.
___
*[`CMakeLists.txt` line 936 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431076873:-Nm1p0UJ2W1LNgsVWDW_:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/CMakeLists.txt#L936)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

As above - I think that the other installations (e.g. spdlog) make sense here because they are used in header files and, therefore, downstream code ends up transitively requiring them, but Catch should only ever appear in translation units for testing, and shouldn't appear in header files that downstream consumers actually see
</blockquote></details>

Done.
___
*[`dependencies/CMakeLists.txt` line 206 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431099083:-Nm1p426DvtiqhXsx2o2:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/dependencies/CMakeLists.txt#L206)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

In contrast to the concerns above, this is fine (you need to acquire it as a dependency for testing purposes). However, the downstream build should be designed to only use `Catch` in test binaries that aren't shipped as part of OpenSim's `package`/`install` target
</blockquote></details>

Done.
___
*[`OpenSim/Actuators/Test/testDeGrooteFregly2016Muscle.cpp` line 33 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431077195:-Nm1p65ZE_nvPvsKf7k5:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/OpenSim/Actuators/Test/testDeGrooteFregly2016Muscle.cpp#L33)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

Noice
</blockquote></details>

Done.
___
*[`OpenSim/Actuators/Test/testMuscles.cpp` line 381 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431080844:-Nm1skNk1gFvo5ym9MPy:b-hvtn4o) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/OpenSim/Actuators/Test/testMuscles.cpp#L381)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

Although it's a little messy, fair enough - a few choice print statements can come in handy when developing/debugging. I'd maybe consider a helper function that doesn't print unless some debug `bool` is enabled, or maybe just writing to the log at `debug`/`trace` level, so that developers an ratchet it up etc. if they want to see the spam
</blockquote></details>

This is definitely temporary, I'd like to come back around to clean this up (perhaps in one of the next phases of the Catch2 transition). When the test is revised we can add proper logging statements as needed.
___
*[`OpenSim/Common/CMakeLists.txt` line 19 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431087832:-Nm1pGfp17tDbmH-Kobe:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/OpenSim/Common/CMakeLists.txt#L19)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

This feels wrong: Catch2 shouldn't be part of any opensim linkage or CMake target creation etc. at all unless it's specifically for the test suite. It'll cause problems on (e.g.) mac/linux, because they lack symbol isolation: any downstream project that happens to also use Catch will end up with multiple-definition related runtime link issues.

(this is already a problem because the newer log implementation requires linking `spdlog`, which links `fmt`, which means OpenSim Creator's (OpenSim-independent engine part) now cannot use `fmt`, because it causes symbol collisions after linking OpenSim).
</blockquote></details>

Done.
___
*[`Vendors/tropter/tests/test_parameter_optimization.cpp` line 19 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431095833:-Nm1tX6K1233djNaIfyg:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/Vendors/tropter/tests/test_parameter_optimization.cpp#L19)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

This file no longer refers to `catch` directly but transitively (can't see it all from here)? If so, it's usually a good idea to explicitly include all used headers (for better isolation).
</blockquote></details>

Done.
___
*[`Vendors/tropter/tests/testing.h` line 36 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431096260:-Nm1sBUD3KYGtyDKusSY:bhtyoxv) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/Vendors/tropter/tests/testing.h#L36)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

Whitespace
</blockquote></details>

This whitespace is added so the backslash for this line lines up with the other backslashes on the right.
___
*[`Vendors/tropter/tests/testing.h` line 52 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431096542:-Nm1sLbI6L0BPq55SBNJ:b61lg4e) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/Vendors/tropter/tests/testing.h#L52)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

Here too
</blockquote></details>

See response above.
___
*[`Vendors/lepton/CMakeLists.txt` line 10 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431092632:-Nm1sOKt4vX6JMToK0sg:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/Vendors/lepton/CMakeLists.txt#L10)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

As above: runtime libraries shouldn't have a dependency on a testing framework - the tests should be separately packaged/linked from the runtime library
</blockquote></details>

Done.
___
*[`Vendors/tropter/tropter/CMakeLists.txt` line 63 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431097319:-Nm1sP1MAdm-Wafretvy:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/Vendors/tropter/tropter/CMakeLists.txt#L63)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

As above - including `Catch2WithMain` as a `PUBLIC` dependency is going to make catch transitively leak over anything using `tropter`
</blockquote></details>

Done.

**Comment by nickbianco** [comment]
Adding @carmichaelong as a reviewer since this touches a lot of files and is worth another pair of eyes.

**Review by carmichaelong — CHANGES_REQUESTED** [review]
Overall looks great. Just only really had a couple of questions (one of which is more for discussion about what isolation is good enough, and I can see both ways)

> Reviewed 52 of 60 files at r1, 3 of 4 files at r2, 13 of 14 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3655#-:-Nm8cuGz0L8dLs48OnOV:b-77mi1)* status: all files reviewed, 13 unresolved discussions (waiting on @adamkewley and @nickbianco)

___
*[`OpenSim/Actuators/Test/testDeGrooteFregly2016Muscle.cpp` line 221 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3655#-Nm8Rykr3zY1sZaYlhNO:-Nm8Rykr3zY1sZaYlhNP:bo6a2s6) ([raw file](https://github.com/opensim-org/opensim-core/blob/2d207233badf9d84fa02f54c3ac6f24f28b0f346/OpenSim/Actuators/Test/testDeGrooteFregly2016Muscle.cpp#L221)):*
> ```cpp
>         CHECK(muscle.calcActiveForceLengthMultiplier(1) == Approx(1));
>         CHECK(muscle.calcForceVelocityMultiplier(-1) ==
>                 Approx(0).margin(1e-10));
> ```

Why was `margin()` added here and not in other calls in this section (especially those that also are compared to `0`?
___
*[`OpenSim/Actuators/Test/testMuscles.cpp` line 381 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431080844:-Nm8SHij-Tt2QnCkJyj2:bsdbmlw) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/OpenSim/Actuators/Test/testMuscles.cpp#L381)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

This is definitely temporary, I'd like to come back around to clean this up (perhaps in one of the next phases of the Catch2 transition). When the test is revised we can add proper logging statements as needed.
</blockquote></details>

This does seem out of scope of this PR. Could consider a quick TODO here to document plans (can be real short, just for a developer to know)
___
*[`OpenSim/Simulation/Test/testProbes.cpp` line 527 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3655#-Nm8YKBqCHoEm03y_uOe:-Nm8YKBqCHoEm03y_uOf:b-zahgtg) ([raw file](https://github.com/opensim-org/opensim-core/blob/2d207233badf9d84fa02f54c3ac6f24f28b0f346/OpenSim/Simulation/Test/testProbes.cpp#L527)):*
> ```cpp
>     model.getMultibodySystem().realize(si, SimTK::Stage::Acceleration);
> 
>     /*double Emuscle0 = */muscWorkProbe->getProbeOutputs(si)(0);
> ```

could consider a TODO here too
___
*[`Vendors/tropter/tests/test_parameter_optimization.cpp` line 19 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431095833:-Nm8_6sY1sBy3elN4CUE:b-f7zl2e) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/Vendors/tropter/tests/test_parameter_optimization.cpp#L19)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Is this also true for all test .cpp files in this folder (since they call include `testing.h` that includes `catch`)? If so, then would be the same for files in `Moco/Test` (that refer to `Testing.h` in that folder)?

Seems like an overarching question is whether inclusion in `testing.h` files is isolated enough, or should it only be included in .cpp files to be clearer?

**Review by carmichaelong — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3655)* status: 68 of 70 files reviewed, 13 unresolved discussions (waiting on @adamkewley and @nickbianco)

___
*[`OpenSim/Actuators/Test/testMuscles.cpp` line 381 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431080844:-Nm8l0Wc9vDmmrYq4Z9d:b-qtxwyo) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/OpenSim/Actuators/Test/testMuscles.cpp#L381)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

This does seem out of scope of this PR. Could consider a quick TODO here to document plans (can be real short, just for a developer to know)
</blockquote></details>

Sorry, I definitely wasn't clear enough! I am OK with the changes that were made to suppress the warnings. I should have been clearer that the TODO I mentioned would be related to adding logging (which could specifically be deferred to a future PR)

**Review by nickbianco — COMMENTED** [review]
Thanks @carmichaelong! Ready for review again.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3655#-:-Nm8jxhKCo2Tx4JWL9gU:b-rflf38)* status: 48 of 70 files reviewed, 13 unresolved discussions (waiting on @adamkewley and @carmichaelong)

___
*[`OpenSim/Actuators/Test/testDeGrooteFregly2016Muscle.cpp` line 221 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3655#-Nm8Rykr3zY1sZaYlhNO:-Nm8gjO44AsBGalNKP6q:b-9jtxu6) ([raw file](https://github.com/opensim-org/opensim-core/blob/2d207233badf9d84fa02f54c3ac6f24f28b0f346/OpenSim/Actuators/Test/testDeGrooteFregly2016Muscle.cpp#L221)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

Why was `margin()` added here and not in other calls in this section (especially those that also are compared to `0`?
</blockquote></details>

For whatever reason, before the changes in this PR, the left hand side was evaluating to exactly zero, while now it is evaluating to ~1e-15. I'm not sure if this is some change in how Catch2 is evaluating a zero comparison, or something else. 
___
*[`Vendors/tropter/tests/test_parameter_optimization.cpp` line 19 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431095833:-Nm8ij1_D1QDUheaPxqU:bv0v964) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/Vendors/tropter/tests/test_parameter_optimization.cpp#L19)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

Is this also true for all test .cpp files in this folder (since they call include `testing.h` that includes `catch`)? If so, then would be the same for files in `Moco/Test` (that refer to `Testing.h` in that folder)?

Seems like an overarching question is whether inclusion in `testing.h` files is isolated enough, or should it only be included in .cpp files to be clearer?
</blockquote></details>

Good point, @carmichaelong. I've added `catch` in all the tropter and Moco tests to match.
___
*[`OpenSim/Actuators/Test/testMuscles.cpp` line 381 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/3655#gh-1431080844:-Nm8hG5M6Ob-fvVhiceU:b-qtjdle) ([raw file](https://github.com/opensim-org/opensim-core/blob/2016811f59d4eccb2ace22c35bf46900177837dd/OpenSim/Actuators/Test/testMuscles.cpp#L381)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

This does seem out of scope of this PR. Could consider a quick TODO here to document plans (can be real short, just for a developer to know)
</blockquote></details>

I have a better fix for this in a newer PR (#3659), so I'll just undo the changes here for now.
___
*[`OpenSim/Simulation/Test/testProbes.cpp` line 527 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3655#-Nm8YKBqCHoEm03y_uOe:-Nm8iYoF3pJFN-EcoInL:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/2d207233badf9d84fa02f54c3ac6f24f28b0f346/OpenSim/Simulation/Test/testProbes.cpp#L527)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

could consider a TODO here too
</blockquote></details>

Done.

File: OpenSim/Common/Test/testAssertion.cpp
Comment (adamkewley) [inline OpenSim/Common/Test/testAssertion.cpp:58]: <3 😉

**Review by adamkewley — COMMENTED** [review]
2nd review: looks fine, the only thing I've noticed is the changing of a tolerance ~~and some whitespace~~

Thanks for fixing the dependency issue. Dependency sprawl is (always has been :wink:) a pet peeve, just because it makes porting the build, integration, etc. harder later on

EDIT: ah, the whitespace is deliberate: sorry, didn't see the comment until now

**Comment by adamkewley** [comment]
> Is this also true for all test .cpp files in this folder (since they call include testing.h that includes catch)? If so, then would be the same for files in Moco/Test (that refer to Testing.h in that folder)?
>
> Seems like an overarching question is whether inclusion in testing.h files is isolated enough, or should it only be included in .cpp files to be clearer?

@carmichaelong , it's a toss-up in C++. I'd say that if a source file uses a symbol from a library/header, it should directly include the library/header, rather than relying on transitive inclusion. It's what projects like `include-what-you-use` push for, because it means that "middle" headers can be changed without breaking other compilation units.

It does, however, increase the number of `include`s in a codebase potentially substantially, if that codebase relies on a lot of otherwise-transtively-included symbols, but I'd argue that it's revealing that the compilation unit is dependent on a large number of other APIs in order to work (which is usually a sign of responsitiblities not quite being hierarchical enough, etc. etc.)

File: OpenSim/Moco/Test/testMocoGoals.cpp
Comment (nickbianco) [reply OpenSim/Moco/Test/testMocoGoals.cpp:470]: @carmichaelong mentioned a different tolerance change, but this change is essentially the same. The `catch` upgrade led to this test barely failing, so I adjusted a bit. I should have commented the reason for it beforehand.

**Comment by nickbianco** [comment]
Thanks @carmichaelong for the review, and thanks @adamkewley for the extra input about header inclusion and isolation. My most recent commit follows this convention, we can put it on our (ever growing) list of things to improve in opensim-core. A good start would be the testing libraries.

If we're all good with the latest changes (including the tolerance adjustment), then I think we can go ahead and merge.

**Comment by nickbianco** [comment]
@adamkewley, pinging again to see if we're good to merge or if you'd like any other changes.

**Comment by nickbianco** [comment]
Thanks @adamkewley!

**Pull request merged**
